[export] Allow shifted constraint ranges in dynamo._export

aakhundov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #114027
* __->__ #114024

Summary: Previously, when we had two dynamic shape symbols `s0` and `s1` bound by the relationship `s1 == s0 + 1`, even when the range constraints were set in accordance with the relationship (e.g., to `[2, 1024]` for `s0` and to `[3, 1025]` for `s1`), `torch._dynamo.export` raised an error saying that the constraint is violated. Here we add a range check between the expression and the constraint and, if the ranges match, don't declare the constraint violated.

We also add a flag to disable the dim constraint solver in `torch._dynamo.export` (not set by default for BC), passed down from the `torch._export.aot_compile`. This is because, even for simple constraints like `s1 == s0 + 1`, the solver claims that the constraint is too complex and the dimension `s0` must be specialized. The new flag is not exposed as a part of the public API (i.e., the one without `_`s in the module names).

Both changes are required to unblock PT2 compilation of an internal model with AOT Inductor.

Test Plan:

```
$ python test/inductor/test_aot_inductor.py -k test_shifted_constraint_ranges
s...
----------------------------------------------------------------------
Ran 4 tests in 53.247s

OK (skipped=1)
```

Reviewers:

Subscribers:

Tasks:

Tags:

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4

aakhundov committed:
Allow shifted constrain ranges in dynamo._export

Summary: Previously, when we had two dynamic shape symbols `s0` and `s1` bound by the relationship `s1 == s0 + 1`, even when the range constraints were set in accordance with the relationship, `torch._dynamo.export` raised the error saying that the constraint is violated. Here we add a range check between the expression and the constraint and, if the ranges match, don't declare the constraint violated.

We also add a flag to disable the dim constraint solver in `torch.dynamo._export` (not set by default for BC). This is because, even for simple constraints like `s1 == s0 + 1`, the solver claims that the constraints are too complex and the dimension `s0` must be specialized.

Both changes are needed to unblock PT2 compilation of an internal model.

Test Plan:

```
$ python test/inductor/test_aot_inductor.py -k test_shifted_constraint_ranges
...
----------------------------------------------------------------------
Ran 4 tests in 53.247s

OK (skipped=1)
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ezyang commented:
Shape env change seems OK, I'll let export peeps bikeshed export

zhxchen17 commented on torch/_export/__init__.py:
```diff
@@ -575,6 +577,7 @@ def _export(
     constraints: Optional[List[Constraint]] = None,
     *,
     preserve_module_call_signature: Tuple[str, ...] = (),
+    disable_constraint_solver: bool = False,
```
@avikchaudhuri @gmagogsfm @aakhundov did we have discussion around this? I'm not recently working on dynamic shape and I don't remember we've talked about letting people disable the solver.

aakhundov replied:
So the issue we're hitting with an internal model export (as part of the internal AOT Inductor lowering) is that the errors like this are generated for constraints like in the unit test added in this PR: `s1 == s0 + 1` (you can see that if you set `disable_constraint_solver=False` in the unit test added in this PR):

```
  - b = L['getattr_10'].size()[0] must be specialized to 32 because the guards generated for it are too complex.
```

Other examples of constraints that cause an error:

    - `Eq(512*(((2*s1)//s0)), 1024)`
    - `512*(((2*s1)//s0)) >= 1024`.

As this is relatively urgent, I was thinking that adding a flag to circumvent the constraint solver to the non-public API (`torch._export`, `torch._dynamo`) would be acceptable as a short-term solution. Which we could then address more fundamentally (like extending the constraint solver to deal with the constraints that lead to errors currently). If you think we should dig into the constraint solver and can find and fix the root cause in a short amount of time, I'm happy to go down that route.

aakhundov committed:
Update on "[export] Allow shifted constraint ranges in dynamo._export"


Summary: Previously, when we had two dynamic shape symbols `s0` and `s1` bound by the relationship `s1 == s0 + 1`, even when the range constraints were set in accordance with the relationship (e.g., to `[2, 1024]` for `s0` and to `[3, 1025]` for `s1`), `torch._dynamo.export` raised an error saying that the constraint is violated. Here we add a range check between the expression and the constraint and, if the ranges match, don't declare the constraint violated.

We also add a flag to disable the dim constraint solver in `torch._dynamo.export` (not set by default for BC), passed down from the `torch._export.aot_compile`. This is because, even for simple constraints like `s1 == s0 + 1`, the solver claims that the constraint is too complex and the dimension `s0` must be specialized. The new flag is not exposed as a part of the public API (i.e., the one without `_`s in the module names).

Both changes are required to unblock PT2 compilation of an internal model with AOT Inductor.

Test Plan:

```
$ python test/inductor/test_aot_inductor.py -k test_shifted_constraint_ranges
s...
----------------------------------------------------------------------
Ran 4 tests in 53.247s

OK (skipped=1)
```

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo ydwu4

[ghstack-poisoned]

aakhundov commented:
As discussed offline with @zhxchen17 and @tugsbayasgalan, removed the `disable_constraint_solver=False` argument from the `torch._export._export`, as not necessary. Here we only care about `torch._export.aot_compile`, which in turn calls `torch._export._export_to_torch_ir`.

avikchaudhuri commented:
> As discussed offline with @zhxchen17 and @tugsbayasgalan, removed the `disable_constraint_solver=False` argument from the `torch._export._export`, as not necessary. Here we only care about `torch._export.aot_compile`, which in turn calls `torch._export._export_to_torch_ir`.

Thanks for not introducing a top-level flag, I'll follow up with what we can do on the constraint solver side.

aakhundov commented:
@pytorchbot merge
